ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
+		 * looking at the first byte? If so, we can run through the string, using
+		 * a table lookup on the first byte to see how many bytes to skip ahead */
+		const unsigned char *mbtab = encoding->mblen_table;
+		unsigned char *p = start, *e = p + byte_offset;
+		size_t len = 0;
+		while (p < e) {
+			p += mbtab[*p];
+			len++;
+		}
+		if (p > e) {
+			len--;
+		}
+		return len;
+	} else {
+		/* Otherwise, do it the slow way: convert the string to wchar, and count those  */
+		size_t len = 0;
+		mbfl_convert_filter *filter = mbfl_convert_filter_new(encoding, &mbfl_encoding_wchar, filter_count_output, 0, &len);
+		mbfl_convert_filter_feed_string(filter, start, byte_offset);
+		mbfl_convert_filter_delete(filter);
+		return len;
+	}
+}
+
+/* Character offset into multi-byte encoded string -> number of bytes up to that position
+ * If requested offset is greater than character length of string, returns byte length of string
+ * Negative offsets count back from the end of the string, but cannot count past the beginning */
+static size_t char_offset_to_byte_offset(const mbfl_encoding *encoding, unsigned char *start, size_t length, ssize_t char_offset)
+{
+	if (char_offset == 0) {
+		return 0;
+	} else if (char_offset < 0) {
+		/* Need to go backwards from end of string */
+		char_offset = -char_offset;
+		if (char_offset >= length) {
+			/* Each character takes _at least_ one byte, so this will definitely go
+			 * to the beginning of the string */
+			return 0;
+		}
+		if (encoding->flag & MBFL_ENCTYPE_SBCS) {
+			return MAX((long)length - char_offset, 0);
+		} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
+			return MAX((long)length - (char_offset << 1), 0);
+		} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
+			return MAX((long)length - (char_offset << 2), 0);
+		} else if (encoding == &mbfl_encoding_utf8) {
+			/* We can easily count backwards from the end of a UTF-8 string
+			 * As we back up, decrement `char_offset` when we find a byte which is the first byte in a character
+			 * It's easy to tell, since any byte which starts with bits 10 is a continuation byte and is not
+			 * the first byte in a character */
+			unsigned char *p = start + length;
+			while (char_offset > 0 && --p > start) {
+				unsigned char c = *p;
+				if (c < 0x80) {
+					char_offset--;
+				} else if ((c & 0xC0) != 0x80) {
+					char_offset--;
+				}
+			}
+			return p - start;
+		} else if (encoding == &mbfl_encoding_utf16le) {
+			/* We can also easily count backwards from the end of a UTF-16 string */
+			length &= -2; /* If there is a 'junk byte' at the end of the string, leave it out */
+			unsigned char *p = start + length;
+			while (char_offset > 0 && (p -= 2) > start) {
+				unsigned char c = *p;
+				if (c < 0xD8 || c >= 0xE0) {
+					char_offset--;
+				} else if ((c & 0xFC) == 0xD8) {
+					char_offset--;
+				}
+			}
+			return p - start;
```

Comment:
Just wondering, why does this optimize UTF16-LE, but not UTF16-BE? Is it more commonly used?

Personally, I would not optimize UTF-16 handling at all. Thankfully, it has very little relevance in PHP's context.

Reply:
Yes, I believe that UTF-16LE is the more common of the two.

If UTF-16LE is not commonly used in PHP, then sure, some of this code could be cut. Can I just think about it a bit more?

Reply:
The typical encodings used with PHP are UTF-8, ASCII and various legacy encodings (latin etc.) I don't think I've ever seen UTF-16 used in the wild.

PHP internally does sometimes use UTF-16 to interact with certain legacy APIs (ICU and Windows), but in these cases the conversion always happens implicitly and is not visible to the user. From the user perspective UTF-8 is provided.

For the encoding of PHP files themselves, even in zend multibyte mode, an ASCII-compatible encoding is required, so UTF-16 cannot be used for that in the first place.

Which is why I don't think optimized UTF-16 implementations are worth it.

Reply:
Understood. Then the UTF-16-specific stuff will disappear. That will cut the code size down a bit more.